### DISCUSSION TITLE
How does the Rest API refresh the cache

### DISCUSSION THREAD

**Discussion by xiaoyv404** [plan-player-analytics/Plan#2904] [Q&A] [answered]
Only the previous data can always be obtained, for example, the request at 12:52 will obtain the data at 12:49

I'm using a translator

**Comment by xiaoyv404** [comment]
My current version is version 5.4 build 1723

**Comment by AuroraLS3** [comment, marked as answer]
Plan uses the json cache to not lock the webserver.

The mechanism for this is same for older version, but what is returned by webserver is a bit different in newer version.

# The update mechanism
![Plan webserver cache](https://user-images.githubusercontent.com/24460436/222880670-a297e85c-e07a-4ca0-82b0-45060441e9da.jpg)

1. Request arrives
2. A new task to update the data in the background is started
3. Data is read from the cache
4. Cached data is returned in Response (200 OK or 304 Not Changed) (_Blue line in the image_)
5. Task finishes in the background
6. Data is written to the cache (_Orange line in image_)

# In 5.5 build 2195 or higher

Timestamp|Where?|What?|Description
---|---|---|---
Cached JSON timestamp|JSON `timestamp`|`{"timestamp": 1677912750123, "timestamp_f": "Today, 06:52", ...}`|Time when Plan updated the data
HTTP Cache timestamp|HTTP response header ETag, request header If-Match|Response: `ETag: "1677912750123"`, Request: `If-Match: "1677912750123"`|Time when browser got new data

### If Browser cache is empty (No ETag)

```
200 OK
Request-Headers
{None}

Response Headers
ETag: 1677912750123

Body
{"timestamp": 1677912750123, "timestamp_f": "Today, 06:52", ...}
```

### If Browser cache has data

If-Match: >= Cached JSON timestamp (We don't have newer data)
```
304 Not Changed
Request-Headers
If-Match: 1677912750123

Response Headers
ETag: 1677912750123

Body
{empty}
```

ETag < Cached JSON timestamp (We have newer data)
```
200 OK
Request-Headers
If-Match: 1677912750123

Response Headers
ETAg: 1677913348120

Body
{"timestamp": 1677913348120, "timestamp_f": "Today, 07:02", ...}
```

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag

# In 5.4 build 1723

```
200 OK

Body
{"timestamp": 1677912750123, "timestamp_f": "Today, 06:52", ...}
```
Later
```
200 OK

Body
{"timestamp": 1677913348120, "timestamp_f": "Today, 07:02", ...}
```

# How to get newest data

- Store time when you start request `var requestStartTime=new Date()`
- Compare request start time to JSON `timestamp`
- If `requestStartTime < json.timestamp`, you have newest data
- Else Request again

**Comment by xiaoyv404** [comment]
Thank you very much for your answer
